File: HomeWorks/HW7/7.1/7.1/list.cpp
#include <stdio.h>
#include "list.h"

struct Node
{
	int value;
	Node *next;
};

struct List
{
	Node *head = nullptr;
};

List *createList()
{
	return new List;
}

bool isEmpty(List *list)
{
	return list->head == nullptr;
}

void addValue(List *list, int value)
{
	Node *current = list->head;
	Node *previous = nullptr;
	while (current != nullptr && current->value < value)
	{
		previous = current;
		current = current->next;
	}
	Node *newNode = new Node {value, current};
	previous == nullptr ? list->head = newNode : previous->next = newNode;
}

bool deleteValue(List *list, int value)
{
	Node *current = list->head;
	Node *previous = nullptr;
	while (current != nullptr && current->value != value)
	{
		previous = current;
		current = current->next;
	}
	if (current == nullptr)
	{
		return false;
	}
	Node *temp = current->next;
	delete current;
	previous == nullptr ? list->head = temp : previous->next = temp;
	return true;
}

void printList(List *list)
{
	Node *current = list->head;
	while (current != nullptr)
	{
		printf("%d ", current->value);
		current = current->next;
	}
}

bool checkSortedList(List *list)
{
	if (!isEmpty(list))
	{
		Node *current = list->head->next;
		Node *previous = list->head;
		while (current != nullptr)
		{
			if (current->value < previous->value)
			{
				return false;
			}
			previous = current;
			current = current->next;
		}
	}
	return true;
}

void deleteList(List *list)
{
	while (!isEmpty(list))
	{
		Node *temp = list->head->next;
		delete list->head;
		list->head = temp;
	}
	delete list;
}

File: Semester2/Homeworks/HW2/Task1/Task1/List.cs
using System;

namespace Task1
{
    class List
    {
        private Node head;

        private class Node
        {
            public Node(int value, Node next)
            {
                this.value = value;
                this.next = next;
            }

            public int value;
            public Node next;
        }

        public int Size { get; private set; }

        public bool IsEmpty { get; private set; }

        public bool AddValue(int value, int position)
        {
            if (position < 0 || position > Size)
            {
                return false;
            }

            Node previous = null;
            var current = head;
            for (var i = 0; i < position; ++i)
            {
                previous = current;
                current = current.next;
            }

            if (previous == null)
            {
                head = new Node(value, head);
            }
            else
            {
                previous.next = new Node(value, current);
            }
            Size++;
            return true;
        }

        public bool DeleteValue(int position)
        {
            if (position < 0 || position > Size - 1)
            {
                return false;
            }

            Node previous = null;
            var current = head;
            for (var i = 0; i < position; ++i)
            {
                previous = current;
                current = current.next;
            }

            if (previous == null)
            {
                head = current.next;
            }
            else
            {
                previous.next = current.next;
            }
            Size--;
            return true;
        }


        public (bool, int) GetValue(int position)
        {
            if (position < 0 || position > Size - 1)
            {
                return (false, 0);
            }

            var current = head;
            for (var i = 0; i < position; ++i)
            {
                current = current.next;
            }
            return (true, current.value);
        }

        public bool SetValue(int value, int position)
        {
            if (position < 0 || position > Size - 1)
            {
                return false;
            }

            var current = head;
            for (var i = 0; i < position; ++i)
            {
                current = current.next;
            }
            current.value = value;
            return true;
        }

        public void Print()
        {
            var current = head;
            for (var i = 0; i < Size; ++i)
            {
                Console.Write($"{current.value} ");
                current = current.next;
            }
        }

        public void Clear()
        {
            head = null;
            Size = 0;
            IsEmpty = true;
        }
    }
}